src/_eval_variables.jl

```diff
@@ -0,0 +1,31 @@
+# Copyright (c) 2025: Joaquim Garcia, Oscar Dowson and contributors
+#
+# Use of this source code is governed by an MIT-style license that can be found
+# in the LICENSE.md file or at https://opensource.org/licenses/MIT.
+
+function _eval_variables end
+
+function _eval_variables(value_fn::Function, t::MOI.ScalarAffineTerm)
+    return t.coefficient * value_fn(t.variable)
+end
+
+function _eval_variables(value_fn::Function, t::MOI.ScalarQuadraticTerm)
+    out = t.coefficient * value_fn(t.variable_1) * value_fn(t.variable_2)
+    return t.variable_1 == t.variable_2 ? out / 2 : out
+end
+
+_eval_variables(value_fn::Function, f::MOI.VariableIndex) = value_fn(f)
+
+function _eval_variables(
+    value_fn::Function,
+    f::MOI.ScalarAffineFunction{T},
+) where {T}
+    # TODO: this conversion exists in JuMP, but not in MOI
+    S = Base.promote_op(value_fn, MOI.VariableIndex)
+    U = Base.promote_op(*, T, S)
+    out = convert(U, f.constant)
```

Comment:
@odow and @blegat 
Is there a special reason why JuMP has this and MOI does not?
Performance?

An alternative to this would be a method for converting ScalarAffineFunction{T} into ScalarAffineFunction{R}.

Reply:
I don't get it, what does not exist in MOI ?

Reply:
This `promote_op` exists in JuMP: https://github.com/jump-dev/JuMP.jl/blob/e5013314bb4bfeac1da63e3cffdc8ac9156e2621/src/aff_expr.jl#L411
but not in MOI: https://github.com/jump-dev/MathOptInterface.jl/blob/ecdf73510d225a28881215af5379bc37e8166820/src/Utilities/functions.jl#L133

Reply:
Ah yes, MOI only works if the return type of the function is the same as the type of `f.constant`. In JuMP, we try to be more general to be more user-friendly.